### PR TITLE
Python: Numerous extensions and fixes

### PR DESCRIPTION
**Pull request by petersteneteg**

### CODE REVIEW

File: modules/python3/interface/pybuffer.cpp [hunk 1]
```diff
@@ -101,26 +104,49 @@ void exposeBuffer(py::module &m) {
...
+                auto data = buffer->getEditableRepresentation<BufferRAM>()->getData();
                 return py::array(pyutil::toNumPyFormat(df), shape, strides, data, py::cast<>(1));
-            }
-        });
+            },
+            [](BufferBase *buffer, py::array data) {
+                auto rep = buffer->getEditableRepresentation<BufferRAM>();
+
+                if (data.dtype() != pyutil::toNumPyFormat(rep->getDataFormat())) {
+                    throw Exception("Invalid data format", IvwContextCustom("Python"));
```
Comment (martinfalk) [inline modules/python3/interface/pybuffer.cpp]: would it be possible to get a more detailed error message?

File: modules/python3/interface/pydataformat.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+void exposeDataFormat(pybind11::module &m) {
+    namespace py = pybind11;
+
+    py::class_<DataFormatBase>(m, "DataFormat")
+        .def_property_readonly("size", &DataFormatBase::getSize)
+        .def_property_readonly("components", &DataFormatBase::getComponents)
+        .def_property_readonly("precision", &DataFormatBase::getPrecision)
+        //.def_property_readonly("numericType", &DataFormatBase::getNumericTypve)
+        //.def_property_readonly("id", &DataFormatBase::getId)
+        .def_property_readonly("max", &DataFormatBase::getMax)
```
Comment (martinfalk) [inline modules/python3/interface/pydataformat.cpp]: is there a reason for not including id and numericType? Probably not really necessary to expose

**Review by martinfalk — APPROVED** [review]
Nice addition to pybind. Good job on the refactoring.

File: modules/python3/interface/pydataformat.cpp (hunk 2 above)
Comment (petersteneteg) [reply modules/python3/interface/pydataformat.cpp]: think the enums are not exposed...

**Review by martinfalk — APPROVED** [review]

**Review by martinfalk — APPROVED** [review]

**Review by martinfalk — APPROVED** [review]
nice

File: modules/python3/pythoninterpreter.cpp [hunk 3]
```diff
@@ -59,11 +57,11 @@ PYBIND11_EMBEDDED_MODULE(inviwo_internal, m) {
...
 }
 
-void PythonInterpreter::init(Python3Module* module) {
+PythonInterpreter::PythonInterpreter(Python3Module* module) : isInit_(false) {
     namespace py = pybind11;
 
     if (isInit_) {
-        throw Exception("Python already initialized", IvwContext);
+        throw ModuleInitException("Python already initialized", IvwContext);
     }
```
Comment (r-englund) [inline modules/python3/pythoninterpreter.cpp]: The ´if (isInit_)´ test should not be needed anymore since this code is now in the constructor.

Comment (petersteneteg) [reply modules/python3/pythoninterpreter.cpp]: Unless someone tries to create annother pythoninterpreter...

**Pull request merged**